tests/kernel/smp/src/main.c

```diff
@@ -717,6 +739,487 @@ void test_workq_on_smp(void)
 		"system workq run on the same core");
 }
 
+static int interrupt_happened[3];
+
+static void t_lock_unlock(void *p1, void *p2, void *p3)
+{
+	int key, k;
+
+	key = irq_lock();
+
+	zassert_equal(_current->base.global_lock_count, 1,
+			"thread global lock cnt %d is incorrect",
+			_current->base.global_lock_count);
+
+	/* tell t2 to trigger interrupt */
+	k_sem_give(&cpuid_sema);
+
+	/* enter sleep, irq lock has to be unlock */
+	k_msleep(100);
+
+	/* wait for t2 done the interrupt */
+	k_sem_take(&cpuid_sema, K_FOREVER);
+
+	/* recursive lock irq */
+	k = irq_lock();
+
+	zassert_equal(_current->base.global_lock_count, 2,
+			"thread global lock cnt %d is incorrect",
+			_current->base.global_lock_count);
+
+
+	/* tell t0 to trigger interrupt*/
+	k_sem_give(&sema);
+
+	/* wait for t0 done the interrupt */
+	k_sem_take(&sema, K_FOREVER);
+
+	irq_unlock(k);
+
+	zassert_equal(_current->base.global_lock_count, 1,
+			"thread global lock cnt %d is incorrect",
+			_current->base.global_lock_count);
+
+	irq_unlock(key);
+
+	zassert_equal(_current->base.global_lock_count, 0,
+			"thread global lock cnt %d is incorrect",
+			_current->base.global_lock_count);
+}
+
+void tIsr_check_entry(void *p1)
+{
+	int t = POINTER_TO_INT(p1);
+
+	interrupt_happened[t]++;
+}
+
+static void t2_do_irq(void *p1, void *p2, void *p3)
+{
+	k_sem_take(&cpuid_sema, K_FOREVER);
+
+	zassert_equal(interrupt_happened[2], 0,
+		"invalid initial int count(%d)", interrupt_happened[2]);
+
+	irq_offload((irq_offload_routine_t)tIsr_check_entry, (void *)2);
+
+	/* check if interrupt still works */
+	zassert_equal(interrupt_happened[2], 1,
+		"invalid initial int count(%d)", interrupt_happened[2]);
+
+	k_sem_give(&cpuid_sema);
+}
+
+/**
+ * @brief Test interrupts remain disabled only while the thread is running
+ *
+ * @ingroup kernel_smp_tests
+ *
+ * @details Validate if the irq_lock() is called by a thread, then it
+ * allows another thread to run (such as giving a semaphore or sleeping
+ * for N milliseconds), the interrupt lock no longer applies and
+ * interrupts may be re-enabled while other processing occurs.
```

Comment:
Yikes, no.  It's true that this has historically worked with irq_lock (because it was recursive, it wasn't possible for the kernel to know whether it was nested or not).  But that was effectively a bug.  Applications that get surprised when the kernel magically releases their locks tend to have terrible race conditions.

We do **not** want to test this condition, there is actually a github issue out there for me to make this an assertion so applications get a proper warning.

Reply:
Appreciate your review ^^, I got it and remove this incorrect test case.